In [122]:
import sys

In [123]:
L =[]
sys.getsizeof(L)

56

In [124]:
L.append(21248)

In [125]:
L

[21248]

In [126]:
sys.getsizeof(L)

88

In [127]:
L.clear()

In [128]:
for i in range(1,50):
    L.append(i)
    print(sys.getsizeof(L))

88
88
88
88
120
120
120
120
184
184
184
184
184
184
184
184
248
248
248
248
248
248
248
248
312
312
312
312
312
312
312
312
376
376
376
376
376
376
376
376
472
472
472
472
472
472
472
472
472


By 32 the size of list gets extended. for empty list size was 56 and added element size increased to 88 and after entering 5 element size increased to 120....

He we are going to create our own List using the ctype module in python.
In python ctype isa  powerful tool that allows us to interact with C data types, and it provides a way of creating new
In Python, the ctypes module provides C compatible data types, and allows calling functions in DLLs/shared libraries. It can be used to wrap these libraries in Pure Python

In [ ]:
import ctypes

In [ ]:
class MyList():
    def __init__(self):
        self.size=1 # Will give how much item can be stored in the array
        self.n=0   # Will keep track of number of items  added so far
        #Create a C type array with size = self.size
        self.A=self.__make_array(self.size)
    
    def __len__(self):
        return self.n

    def __make_array(self,capacity):
        return (capacity*ctypes.py_object)()


In [ ]:
L=MyList()

In [ ]:
type(L)

In [ ]:
len(L)

While adding the append fuctionality we need to check two things:
    (i)Whether the list is vacant, append the element provide
        if self.n==self.size:
    (ii)If it's not empty, Need to resize the array

In [ ]:
class MyList():
    def __init__(self):
        self.size=1 # Will give how much item can be stored in the array
        self.n=0   # Will keep track of number of items  added so far
        #Create a C type array with size = self.size
        self.A=self.__make_array(self.size)
    
    def __len__(self):
        return self.n

    def __make_array(self,capacity):
        return (capacity*ctypes.py_object)()
    
    def append(self,item):
        #check for the  availability of memory first
        if self.n==self.size:
            self.__resize(self.size*2)
        #Adding an element to the list
        self.A[self.n]=item
        self.n+=1
    
    def __resize(self,newCapacity):
        #Creating new Array of given capacity 
        new_array=self.__make_array(newCapacity)
        self.size=newCapacity
        #copy the content of A to new_array
        for i in range(self.n):
            new_array[i] = self.A[i]
        #reassign A
        self.A = new_array

    

    #to print the list
    def __str__(self):
        result=''
        for i in range(self.n):
            result=result+str(self.A[i])+','
        return '['+result[:-1]+']'
    
    #to provide the indexing
    def __getitem__(self,index):
        if index<0 or index>=self.n :
            return "IndexError:list index out of range"
        else:
            return self.A[index]
    
    def pop(self):
        



In [ ]:
L=MyList()

In [ ]:
L.append('Sooraj')

In [ ]:
len(L)

In [ ]:
print(L)

In [ ]:
L[1]

In [ ]:
stock=[298,305,320,301,292]

In [ ]:
for index,value in enumerate(stock):
    print(f"Day {index+1} : Stock Price {value}")

# Time complexity
1.accessing the element,stock[2] is O(1) because accessing an array by index directly takes constant time. 

2.Checking for a stock price to find the day
    for i range(len(stock)):
        if price[i]==301
        return i 
    // This loop will take O(n) time as it checks each element of 'price' list with the given value and returns the first occurrence

3.Print all the prices
    for price range(len(stock)):
        print(price)
    complexity of this is also  O(n), where n is the number of elements in the list "price".

4.Insert a new price at given index
    stock.insert(index,value)
    This operation will take O(n) time as it needs to shift all the elements from that position till end of the list one place forward.

5.Deleting an element
    stock.remove(value)
    will take a time complexity of O(n). In worst case scenario it has to traverse through all the elements until it finds the value that

6.Sorting the list
    The Python built-in function sorted() has a time complexity of O(n log(n)).


Static array and Dynamic Array(Geometric  Progression)

In [ ]:
expense={'January':2200,'February':2350,'March':1897,'April':2130,'May':2190}
# print(expense['January'])
# expense['February']-expense['January']
for key,value in expense.items():
    print(value)

In [ ]:
expense=[2200,2350,2600,2130,2190]

In [ ]:
expense[1]-expense[0]

In [ ]:
sum=0
for i in range(3):
    sum=sum+expense[i]
print(sum)

In [ ]:
search=2000
for i in range(len(expense)):
    if expense[i]==2000:
        break
        print("was found at")
    else:
        print("Not Found")
        break

In [ ]:
expense.append(1980)

In [ ]:
refund=expense[3]-200
expense.insert(3,refund)

In [ ]:
expense

In [ ]:

heros=['spider man','thor','hulk','iron man','captain america']

In [ ]:
heros.append('black panther')

In [ ]:
heros

In [ ]:
heros.pop()


In [ ]:
heros.insert(3,'black panther')

In [ ]:
heros.sort()

In [ ]:
dir(list)

In [ ]:
max_number=int(input("Enter a number"))
for i in range(1,max_number+1):
    if (i%2!=0):
        print(i)


In [120]:
max_number=int(input("Enter a number"))
odd_numbers=[]
for i in range (1,max_number+1):
    if i%2!=0:
        odd_numbers.append(i)
print ("The list of odd numbers from  1 to",max_number,"is:")
print (odd_numbers)

The list of odd numbers from  1 to 100 is:
[1, 3, 5, 7, 9, 11, 13, 15, 17, 19, 21, 23, 25, 27, 29, 31, 33, 35, 37, 39, 41, 43, 45, 47, 49, 51, 53, 55, 57, 59, 61, 63, 65, 67, 69, 71, 73, 75, 77, 79, 81, 83, 85, 87, 89, 91, 93, 95, 97, 99]


Linked List has two main advantage  over Array. 
1) You dont need to pre-allocate space
    It is dynamic and can grow or shrink in size as required, unlike an array which needs to be declared with a fixed size at the time of creation.
2) Elements can be added or removed dynamically or insertion is easier

Inserting element at beginning- O(1)
Insertion in the middle - O(n)
deleting element at beginning-O(1)
Deleting an element from end of list - O(n)
Linked List Traversal-O(n)
Accessing Element by Value-O(n)

Doubly-Linked list have the pointer to the previous node and to the next node
With doubly linkedlist , we can traverse backward as well as forward and backward traveesal becomes easier



In [57]:
class Node:
    def __init__(self, data=None, next=None):
        self.data = data #data in the node
        self.next = next  #pointer to the next node

class LinkedList:
    def __init__(self):
        self.head = None
    
    def insert_at_begining(self,data):
        node=Node(data,self.head) #create the data and  point it at head
        self.head=node             #set new node as head of linked list

    def insert_at_end(self,data):
        if self.head is None:
            self.head=Node(data,None)
            return
        itr=self.head
        while itr.next:
            itr=itr.next
        
        itr.next=Node(data,None)

    def insert_values(self,data_list):
        self.head=None
        for data in data_list:
            self.insert_at_end(data)

    def get_length(self):
        count=0
        itr=self.head
        while itr:
            count+=1
            itr=itr.next
        return count
    
    def remove_at(self,index):
        if index<0 or index>=self.get_length():
            return "Index Error- Index is out of range or invalid index"
        
        if  index==0:                   #removing from beginning
            self.head=self.head.next
            return
        
        count=0
        itr=self.head
        while itr:
            if count==index-1:
                itr.next=itr.next.next
                break
            itr=itr.next
            count+=1
    def insert_at(self,index,data):
        if index<0 or index>=self.get_length():
            return "Index Error- Index is out of range or invalid index"
        if index==0:
            self.insert_at_begining(data)
            return
        
        count=0
        itr=self.head
        while itr:
            if count==index-1:
                node=Node(data,itr.next)
                itr.next=node
                break
            itr=itr.next
            count+=1


    def display(self):
        if self.head is None:
            return ("The list is empty")
        itr=self.head #create an iteratpr variable itr and assigned head to it
        llstr=''
        while itr:
            llstr=llstr+str(itr.data)+"-> "
            itr=itr.next
        
        print(llstr)





In [16]:
ll=LinkedList()
ll.insert_at_begining(10)
ll.insert_at_begining(110)

In [34]:
ll.display()

Sooraj-> Chandini-> 10-> 12-> 


In [18]:
ll.insert_at_end(2015)

In [58]:
ll=LinkedList()
ll.insert_values(['Sooraj','Chandini',10,12])

In [59]:
ll.get_length()

4

In [60]:
ll.display()

Sooraj-> Chandini-> 10-> 12-> 


In [56]:
ll.remove_at(-1)

'Index Error- Index is out of range or invalid index'

In [64]:
ll.display()

JackFruit-> Sooraj-> Chandini-> Mango-> 10-> 12-> 


In [63]:
ll.insert_at(0,'JackFruit')

In [ ]:
class Node:
    def __init__(self, data=None, next=None):
        self.data = data #data in the node
        self.next = next  #pointer to the next node

class LinkedList:
    def __init__(self):
        self.head = None
    
    def insert_at_begining(self,data):
        node=Node(data,self.head) #create the data and  point it at head
        self.head=node             #set new node as head of linked list

    def insert_at_end(self,data):
        if self.head is None:
            self.head=Node(data,None)
            return
        itr=self.head
        while itr.next:
            itr=itr.next
        
        itr.next=Node(data,None)

    def insert_values(self,data_list):
        self.head=None
        for data in data_list:
            self.insert_at_end(data)

    def get_length(self):
        count=0
        itr=self.head
        while itr:
            count+=1
            itr=itr.next
        return count
    
    def remove_at(self,index):
        if index<0 or index>=self.get_length():
            return "Index Error- Index is out of range or invalid index"
        
        if  index==0:                   #removing from beginning
            self.head=self.head.next
            return
        
        count=0
        itr=self.head
        while itr:
            if count==index-1:
                itr.next=itr.next.next
                break
            itr=itr.next
            count+=1
    def insert_at(self,index,data):
        if index<0 or index>=self.get_length():
            return "Index Error- Index is out of range or invalid index"
        if index==0:
            self.insert_at_begining(data)
            return
        
        count=0
        itr=self.head
        while itr:
            if count==index-1:
                node=Node(data,itr.next)
                itr.next=node
                break
            itr=itr.next
            count+=1

    def insert_value_after(self,data_after,data_to_insert):
        

    def display(self):
        if self.head is None:
            return ("The list is empty")
        itr=self.head #create an iteratpr variable itr and assigned head to it
        llstr=''
        while itr:
            llstr=llstr+str(itr.data)+"-> "
            itr=itr.next
        
        print(llstr)

In [80]:
# Hash table

def get_hash(key):
    h=0
    for char in key:
        h=h+ord(char)
    return h%30

In [81]:
get_hash('March 6')

7

In [82]:
get_hash('March 13')

23

In [130]:
class HashTable():
    def __init__(self):
        self.MAX=31
        self.arr=[None for i in range(self.MAX)]
    
    def get_hash(self,key):
        h=0
        for char in key:
            h=h+ord(char)
        return h%self.MAX
        
    def __setitem__(self,key,data):
        h=self.get_hash(key)
        self.arr[h]=data

    def __getitem__(self,key):
        h=self.get_hash(key)
        return self.arr[h]
    
    def __delitem__(self,key):
        h=self.get_hash(key)
        self.arr[h]=None
    

In [126]:
h=HashTable()

In [127]:
h['March 9']=200

In [128]:
h['March 9']

200

In [ ]:
#to Avoid collision
1> Linear probing: Checks for the empty space linearly and save the data


In [133]:
class HashTable():
    def __init__(self):
        self.MAX=31
        self.arr=[[] for i in range(self.MAX)]
    
    def get_hash(self,key):
        h=0
        for char in key:
            h=h+ord(char)
        return h%self.MAX
        
    def __setitem__(self,key,data):
        h=self.get_hash(key)
        found=False
        for index,element in enumerate(self.arr[h]):
            if len(element)==2 and element[0]==key:
                self.arr[h][index]=(key,data)
                found=True
                break
            if not found:
                self.arr[h].append((key,data))
         

    def __getitem__(self,key):
        h=self.get_hash(key)
        for element in self.arr[h]:
            if element[0]==key:
                return element[1]
    
    def __delitem__(self,key):
        h=self.get_hash(key)
        for index,element in enumerate(self.arr[h]):
            if  element[0]==key:
                del self.arr[h][index]
    

# Stack-LIFO
#Push/Pop operations are performed at the top of the stack.
# with a time complexity of O(1)
# Search element by value
# The space complexity is O(n), where n is the number of elements in the stack.

Use cases:
    Fuction calling in any programming  language is managed using stack.
    undo functionality in any editor uses stack to track down last set of operation

In [134]:
from collections import deque
stack=deque()

In [159]:
class Stack():
    def __init__(self):
        self.container=deque()
    
    def push(self,item):
        self.container.append(item)
        return f"Item {item} pushed to stack"
        
    def pop(self):
        try:
            return f"{self.container.pop()} popped from the stack"
        except IndexError:  
            return "Stack is empty!"
        
    def peek(self):
        if len(self.container)>0:
            return self.container[-1]
        else:
            return 'The stack is empty!'
    
    def is_empty(self):
        if len(self.container)==0:
            return "Stack is empty"
        else:
           return "Stack is not empty"

    def size(self):
        return f"The length of the stack is {len(self.container)}"
    
    def display(self):
        reversed_container=list(reversed(self.container))
        for items in reversed_container:
            print(items)
        


In [154]:
s=Stack()
s.push(1)
s.push(2)

'Item 2 pushed to stack'

In [155]:
s.size()

'The length of the stack is 2'

In [158]:
s.display()

1


In [157]:
s.pop()

'2 popped from the stack'

In [161]:
s.is_empty()